pkg/runtime.v2/framework/plugins/jobset/jobset.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jobset
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	batchv1 "k8s.io/api/batch/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	jobsetv1alpha2 "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+	runtime "github.com/kubeflow/training-operator/pkg/runtime.v2"
+	"github.com/kubeflow/training-operator/pkg/runtime.v2/framework"
+)
+
+type JobSet struct {
+	client     client.Client
+	restMapper meta.RESTMapper
+	scheme     *apiruntime.Scheme
+}
+
+var _ framework.WatchExtensionPlugin = (*JobSet)(nil)
+var _ framework.ComponentBuilderPlugin = (*JobSet)(nil)
+
+const Name = "JobSet"
+
+func New(_ context.Context, c client.Client, _ client.FieldIndexer) (framework.Plugin, error) {
+	return &JobSet{
+		client:     c,
+		restMapper: c.RESTMapper(),
+		scheme:     c.Scheme(),
+	}, nil
+}
+
+func (j *JobSet) Name() string {
+	return Name
+}
+
+func (j *JobSet) Build(ctx context.Context, info *runtime.Info, trainJob *kubeflowv2.TrainJob) (client.Object, error) {
+	if info == nil || info.Obj == nil || trainJob == nil {
+		return nil, fmt.Errorf("runtime info or object is missing")
+	}
+	raw, ok := info.Obj.(*jobsetv1alpha2.JobSet)
+	if !ok {
+		return nil, nil
+	}
+	jobSetBuilder := NewBuilder(client.ObjectKeyFromObject(trainJob), kubeflowv2.JobSetTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      info.Labels,
+			Annotations: info.Annotations,
+		},
+		Spec: raw.Spec,
+	})
+	jobSet := jobSetBuilder.
+		ContainerImage(trainJob.Spec.Trainer.Image).
```

Comment:
We need to make sure we insert image only for replicated Job with training script (e.g. it should have Node name).

Reply:
Yeah, actually, we need to implement many build parts here.
But, I would like to focus on the interface and a few example implementations.

Reply:
@tenzen-y Let's add the TODO mark for future developers.

Reply:
SGTM.

Reply:
Done.